dnf/cli/commands/history.py

```diff
@@ -219,24 +243,140 @@ def str2transaction_id(s):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "rpms" in "{fn}", object expected.'.format(fn=fn))
+            )
+
+        groups = self.replay_data.get("groups", {})
+        if not isinstance(groups, dict):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "groups" in "{fn}", object expected.'.format(fn=fn))
+            )
 
-        tids, merged_tids = self._args2transaction_ids()
+        environments = self.replay_data.get("environments", {})
+        if not isinstance(groups, dict):
+            raise dnf.exceptions.Error(
+                _('Unexpected type of "environments" in "{fn}", object expected.'.format(fn=fn))
+            )
 
+        self._na_reason_cache = {}
+
+        for na, pkg_data in rpms.items():
+            subj = hawkey.Subject(na)
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                raise dnf.exceptions.Error(_('Cannot parse name.arch for package "{na}".').format(na=na))
+
+            parsed_na = nas[0]
+            try:
+                action = pkg_data["action"]
+
+                # validate the reasons before running the transaction and cache them
+                try:
+                    self._na_reason_cache[na] = \
+                        libdnf.transaction.StringToTransactionItemReason(pkg_data["reason"])
+                except IndexError as e:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package reason "{reason}" for package "{pkg}" in "{fn}".')
+                            .format(reason=pkg_data["reason"], pkg=na, fn=fn)
+                    )
+
+                if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["to"], parsed_na.arch)
+                    if action in ("Upgrade", "Downgrade", "Reinstall"):
+                        from_nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                        query = dnf.subject.Subject(from_nevra).get_best_solution(self.base.sack)["query"]
+                        if not query.installed():
+                            raise dnf.exceptions.Error(
+                                _('Package "{pkg}" not found installed for action "{action}" in "{fn}".')
+                                    .format(pkg=from_nevra, action=action, fn=fn)
+                            )
+
+                elif action == "Removed":
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                else:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package action "{action}" for package "{pkg}" in "{fn}".')
+                            .format(action=action, pkg=na, fn=fn)
+                    )
+
+                query = dnf.subject.Subject(nevra).get_best_solution(self.base.sack)["query"]
```

Comment:
Please could you add following keywords with_nevra=True, with_provides=False, with_filenames=False. Alse please could you specify forms=with a proper form?